Revert a couple of commits that are not in the Apache Kafka 3.4 branch

ijuma opened the pull request:
The right branching point should have been 96b1db510a32b99.
I incorrectly picked the branching point to be the commit before
the version bump, but there was a gap between the branching
point and the version bump merge in AK trunk.

Most of the additional commits were cherry-picked to 3.4, but
two were not. I revert them both in this PR. I will push the
reverts directly once this PR is approved.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

gbadoni reviewed (approved):
Following are the commits to AK between 96b1db5 and version bump to 3.5. As long as all others have been cherry picked to 3.4 except 2, looks good to me. 

KAFKA-14415: Faster ThreadCache (apache#12903) 

KAFKA-14260: add synchronized to prefixScan method (apache#12893) 

KAFKA-14352: Rack-aware consumer partition assignment protocol change… 

MINOR: Fix some compiler warnings (apache#12912) 

MINOR: Move dynamic config logic to DynamicConfigPublisher (apache#12958 

KAFKA-14367; Add ListGroups to the new GroupCoordinator interface (… 

KAFKA-14443: Close topic creation Admin clients in MM2 connectors (ap… 

KAFKA-14432: RocksDBStore relies on finalizers to not leak memory (ap… 

MINOR: Bump trunk to 3.5.0-SNAPSHOT (apache#12960)

ijuma commented:
> KAFKA-14415: Faster ThreadCache ([apache#12903](https://github.com/apache/kafka/pull/12903))

https://github.com/apache/kafka/commit/96b1db510a32b99b57c27a9b3001c9e15af722d4

> KAFKA-14260: add synchronized to prefixScan method ([apache#12893](https://github.com/apache/kafka/pull/12893))

https://github.com/apache/kafka/commit/f292e1960b931922a7a4be19424fd1e4580d9264

> KAFKA-14352: Rack-aware consumer partition assignment protocol change…

https://github.com/apache/kafka/commit/f5f1060a81dd200e573ef69574ea75b64b73474c

> MINOR: Fix some compiler warnings ([apache#12912](https://github.com/apache/kafka/pull/12912))

Reverted in this PR.

> MINOR: Move dynamic config logic to DynamicConfigPublisher ([apache#12958](https://github.com/apache/kafka/pull/12958)

https://github.com/apache/kafka/commit/e94a25173dc469da650f4d620ab312a1363242a2

> KAFKA-14367; Add ListGroups to the new GroupCoordinator interface (…

Reverted this PR.

> KAFKA-14443: Close topic creation Admin clients in MM2 connectors (ap…

https://github.com/apache/kafka/commit/4ab0c54a5d872add78a0cb20720fd33fa5469c4d

> KAFKA-14432: RocksDBStore relies on finalizers to not leak memory (ap…

https://github.com/apache/kafka/commit/8c1e04e42873aa68753bbc34240624219922cc39

> MINOR: Bump trunk to 3.5.0-SNAPSHOT ([apache#12960](https://github.com/apache/kafka/pull/12960))

This should not be in the 3.4 branch.

ijuma commented:
Pushed commits to https://github.com/confluentinc/kafka/tree/3.4.
